Merge dev progress into main

henning-oden opened the pull request:
There is still a lot of work left to do, including refactoring of some business logic that should be in a service rather than a controller.

henning-oden committed:
Initial setup w. unsuccessful user test

henning-oden committed:
Fixes #10 by putting UserDetailsService in a separate class

henning-oden committed:
CustomUserDetailsService throws NullPointerException at line 42. Commit to save user progress.

henning-oden committed:
User-related files moved to backup to allow progress

henning-oden committed:
Implemented SystemUser

henning-oden committed:
Merge remote-tracking branch 'origin/dev' into dev

# Conflicts:
#	StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/SystemUser.java
#	StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/services/CustomUserDetailsService.java

henning-oden committed:
Revert "User-related files moved to backup to allow progress"

This reverts commit f74f1238

henning-oden committed:
Working signin in memory

henning-oden committed:
Attempted H2 in-memory database implementation. Runtime errors exist.

henning-oden committed:
Cleaned up config.

henning-oden committed:
Reverted to MySQL and no Liquibase.

CustomUserDetailsService inserts an entry with all columns set to null into the database rather than supplying data.

henning-oden committed:
Successful database with MySQL and Liquibase

Reading documentation and iteratively trying things finally lead to success. The test user needs to be created outside of the program which is easily done with MySQL workbench. Another method that will be implemented in the program is a data seed service. Properties and controls to ensure that isn't run in prod - or just removing the seeder completely from the program before release - are a later priority.

henning-oden committed:
Reset web security config

henning-oden committed:
First seeding implementation.

Needs overhaul / removing at later stage.

henning-oden committed:
Stock class implemented.

henning-oden committed:
Retroactive inclusion of Stock class

henning-oden committed:
Refactored stock deletion test

to ensure it fails when there is a problem deleting the stock

henning-oden committed:
Seed stocks

henning-oden committed:
Refactoring and future proofing of entities

henning-oden committed:
StockWatch class implemented

henning-oden committed:
StockWatchService refactored

Uses functional style of programming and improved null safety.

henning-oden committed:
Implemented simple get endpoint

henning-oden committed:
First steps with Alpaca API

henning-oden committed:
Using Lombok (correct provided version)

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/Application.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.context.annotation.ComponentScan;
+
+@SpringBootApplication
+@ComponentScan(basePackages = "com.henning.oden.java.StockAlert.Backend")
```
Do we need this?

henning-oden replied:
It doesn't seem like it.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/controllers/AuthController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @PostMapping("/signin")
+    public ResponseEntity<Map<Object, Object>> signin(@RequestBody AuthenticationRequest data) {
```
Use DTO (transfer objects) instead

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/controllers/AuthController.java:
```diff
@@ -0,0 +1,61 @@
+    public ResponseEntity<Map<Object, Object>> signin(@RequestBody AuthenticationRequest data) {
+        try {
+            String username = data.getUsername();
+            authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(username, data.getPassword()));
```
move logic to service method

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/AuthenticationRequest.java:
```diff
@@ -0,0 +1,18 @@
+
+@AllArgsConstructor
+@NoArgsConstructor
+public class AuthenticationRequest implements Serializable {
```
this is DTO object, move to separate DTO folder

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/AuthenticationRequest.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.io.Serializable;
+
+@AllArgsConstructor
```
Instead of all these annotations, can you use @Value to make it immutable. Strive for immutability in general

henning-oden replied:
Good idea! Thanks for pointing this option out.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/Stock.java:
```diff
@@ -0,0 +1,33 @@
+package com.henning.oden.java.StockAlert.Backend.entities;
+
+import lombok.*;
```
no * imports pls

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/StockWatch.java:
```diff
@@ -0,0 +1,72 @@
+    private double minPrice;
+
+    @Column
+    @Getter @Setter
```
do we need such annotations on fields?

henning-oden replied:
Without them JPA seems to rely on calling the no arg constructor which would then leave the fields uninitialized when fetching entities of this type from the database.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/StockWatch.java:
```diff
@@ -0,0 +1,72 @@
+        this.stockId = stockId;
+        this.minPrice = minPrice;
+        this.maxPrice = maxPrice;
+        this.timesExceeded = countExceeded < 1 ? 5 : countExceeded;
```
not sure why we need these here

henning-oden replied:
Leftover from .NET Core conventions that don't apply in the same way with Spring.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/StockWatch.java:
```diff
@@ -0,0 +1,72 @@
+        this.alertThreshold = alertThreshold < 1 ? 5 : alertThreshold;
+    }
+
+    public StockWatch(double minPrice, double maxPrice, int countExceeded, int alertThreshold, Stock stock, SystemUser user) {
```
isnt this @allargs from lombok?

henning-oden replied:
No, because this one does not set the id.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/repos/StockWatchRepository.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.Optional;
+
+public interface StockWatchRepository extends JpaRepository<StockWatch, Long> {
+    Optional<Collection<StockWatch>> findByStockId(long stockId);
```
Use List

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/services/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Collections;
+
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
```
Do we need interface?

henning-oden replied:
I tried to remove the interface and the next time I ran the application and tried to sign in, I got a StackOverflowError. Nothing else was changed other than remove the interface and override.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/services/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
```
Use Optional

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/services/PasswordEncoderService.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class PasswordEncoderService {
```
May be move to another service?

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/services/StockService.java:
```diff
@@ -0,0 +1,32 @@
+        this.stockRepository = stockRepository;
+    }
+
+    public Stock findStockByCode(String code) throws NotFoundException {
```
Use Optional

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/services/StockWatchService.java:
```diff
@@ -0,0 +1,46 @@
+        stockWatches = watches;
+    }
+
+    public Collection<StockWatch> findStockWatchesByStock(Stock stock) {
```
Use List

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/util/dev/DataSeeder.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.Collections;
+
+@Component
+public class DataSeeder {
```
DO it in liquibase and remove this class

henning-oden replied:
Attempt to do this with Liquibase fails as the password is encrypted with BCrypt and login seems to become impossible after seeding with Liquibase.

svetli-n replied:
Not sure I understand the probl. 

You encrypt, and use the result  string as value, insert ok but then when compares on login doesn't work?

henning-oden replied:
Yes. I might have missed a letter somewhere, so I am going to try again

henning-oden replied:
I have checked to make sure I didn't miss any characters and for some reason, login just fails with the inserted password hash. Doing everything over from the start with the SeedData class implementation works, not the Liquibase approach. I have confirmed that the entry in the database is the same in both cases.

svetli-n commented on StockWatch.Backend/src/main/resources/alpaca.properties:
```diff
@@ -0,0 +1,5 @@
+key_id = PKHP14LN183KRWGJMLKD
```
use env vars

henning-oden replied:
Solved partially; Alpaca doesn't seem to accept environment variables for these. An issue will be posted on the official repository. The rest of the program does use environment variables now and profiles will be implemented at a later point.

svetli-n replied:
I mean like https://stackoverflow.com/questions/35531661/using-env-variable-in-spring-boots-application-properties

svetli-n replied:
Or see this

String keyID = "<some key ID>";
String secret = "<some secret>";
AlpacaAPI alpacaAPI = new AlpacaAPI(keyID, secret);

you can load them from spring app props?

henning-oden replied:
I haven't tried from the spring app properties just yet but from the variables themselves. That might be the issue; will try the approach of going through the main Spring properties when I continue.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/controllers/AuthController.java:
```diff
@@ -0,0 +1,61 @@
+            String token = jwtTokenProvider.createToken(username, this.userDetailsService.loadUserByUsername(username)
+                    .getAuthorities().stream().map(Object::toString).collect(Collectors.toList()));
+
+            Map<Object, Object> model = new HashMap<>();
```
Use Map.of

henning-oden replied:
New feature I had no knowledge of.

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/entities/StockWatch.java:
```diff
@@ -0,0 +1,72 @@
+    @Getter @Setter
+    private int alertThreshold;
+
+    @JsonInclude()
```
Why is that needed?

svetli-n commented on StockWatch.Backend/src/main/java/com/henning/oden/java/StockAlert/Backend/services/StockService.java:
```diff
@@ -0,0 +1,32 @@
+        return stockRepository.findByCode(code).orElseThrow(() -> new NotFoundException("Stock with code " + code + " not found"));
+    }
+
+    public Stock findStockByCommonName(String commonName) throws NotFoundException {
```
avoid throwing checked exeptions

henning-oden committed:
.

henning-oden committed:
.

henning-oden committed:
Use environment variables

Alpaca API might not be able to use environment variables for initialization. Also, Environment Profiles at a later stage.

henning-oden committed:
Refactor AuthController

Business logic extracted from controller. Authentication with Jwt changed extensively as a consequence.

henning-oden committed:
Minor code optimizations

henning-oden committed:
Replaced DataSeeder with Liquibase insert

This is still not the best solution to seeding data, though this is what I have been able to make work so far. Ideally, loadUpdateData in Liquibase should be used, though for some reason I was getting NullPointerExceptions when I tried that method. More thorough troubleshooting to follow.

henning-oden committed:
Replaced DataSeeder with Liquibase insert

This is still not the best solution to seeding data, though this is what I have been able to make work so far. Ideally, loadUpdateData in Liquibase should be used, though for some reason I was getting NullPointerExceptions when I tried that method. More thorough troubleshooting to follow.

henning-oden committed:
More consistent usage of Optional

Also removed some checked exception throw declarations.

henning-oden committed:
Revert "Replaced DataSeeder with Liquibase insert"

This reverts commit 1924adbd

henning-oden committed:
Fix unexpected NPE in SystemUser during app start with empty database

henning-oden committed:
Removed imports of full packages (* imports)

henning-oden committed:
Removed @ComponentScan annotation

henning-oden committed:
Environment variables for Alpaca

Accomplished through application.properties. Also necessitated refactoring or removal of test for configuration which was of little use either way. Removal was chosen.

henning-oden committed:
Cleaner implementation of JwtAuthentication.

henning-oden committed:
Refactor database seeding

henning-oden committed:
Refactor Auth controller

Not using ResponseEntity wrapper anymore, and new DTO for the /auth/info endpoint.

henning-oden committed:
Actually fetching stock data from Alpaca now

henning-oden committed:
Store price data in database

Price values end up saved as integers for some reason, it seems. This must be investigated.

henning-oden committed:
Fix Microsoft stock code in dbchangelog.xml

henning-oden committed:
Email address added to users

Includes validation via token sent by e-mail.

henning-oden committed:
Use BigDecimal for non-integers in program

henning-oden committed:
Use decimal SQL type correctly in dbchangelog.xml

henning-oden committed:
Remove time adjustment for testing

henning-oden committed:
Refactor constructors and StocksController

Using builder for StockWatch and a static factory method for SystemUser.

henning-oden committed:
Minor refactoring

henning-oden committed:
Refactor EmailConfirmationTokenService

Better use of optional. In verifyUserEmail the use of Optional results in arguably more stylish code, though at this point performance would probably benefit from omitting Optional altogether.

henning-oden committed:
Refactor creation of StockWatch.

henning-oden committed:
Minor refactoring of StockPriceDataService

henning-oden committed:
Added email notofications. Basic backend feature complete.

Still much room for improvement and further tests to be implemented.

henning-oden committed:
Effective Java Item 12: Always override toString

Temporary but still functional overrides of toString() have been implemented in all implementation classes.

henning-oden committed:
Minor fixes

henning-oden committed:
Read and Delete endpoints for Stock watches.

henning-oden committed:
Reset times exceeded when price interval is not exceeded

henning-oden committed:
Basic React frontend (todo functionality)

henning-oden committed:
Constant functions, start usingContext

henning-oden committed:
Basic navigation

henning-oden committed:
CORS + attempt two props in one context

henning-oden committed:
working login?

henning-oden committed:
Partial progress list stock watches

henning-oden committed:
useEffect for data fetching

henning-oden committed:
Working but no re-render on fetch

henning-oden committed:
Progress with Registration Issues

henning-oden committed:
Stringify body

henning-oden committed:
Minor tidying

henning-oden committed:
Create watches

henning-oden committed:
Delete watches

henning-oden committed:
Refactor create watches

henning-oden committed:
Rename SelectedStockContext

henning-oden committed:
Edit stock watches (no prepop. form due cors iss.)

henning-oden committed:
Fix prepopulation of stock watch data for edit

henning-oden committed:
Minor visual tweaks

henning-oden committed:
Fix number inputs

henning-oden committed:
Very simple basic error handling

henning-oden committed:
Better REST-compliance in API.

henning-oden committed:
More graceful error handling

henning-oden committed:
More consistent objects in frontend errors

henning-oden committed:
Further improved REST compliance.

henning-oden committed:
Attempt to handle JwtException

henning-oden committed:
Revert "Attempt to handle JwtException"

This reverts commit 2fcd6465
